CFC_DataCollector/recommender/alternative_trips_module.py

```diff
@@ -0,0 +1,105 @@
+
+
+"""
+
+
+# Invoked in recommendation pipeline to get perturbed trips user should consider
+def calc_alternative_trips(trip_iterator):
+    while (trip_iterator.next() != None):
+        curr_trip = trip_iterator.next()
+        if not curr_trip.getPipelineFlags().alternativesStarted:
+            curr_trip.getPipelineFlags().startAlternatives()
+            curr_unique_id = curr_trip._id
+            list_of_perturbed_trips = find_perturbed_trips(curr_unique_id)
+            
+            #@TODO: need a method that initializes the relationship bewteen a trip, and its perturbations
+            #initialize_perturbations(curr_unique_id)
+            schedule_queries(list_of_perturbed_trips)
+
+
+
+#this call should be made asynchronously
+#if it returns true, that means the perturbation queries were finished, and we are now able to move on to the next pipeline
+#if it returns false, that means the perturbation queries were not finished
+def invoke_next_pipeline(_id):
+    if check_all_queries_made(_id) == True:
+        #invoke next pipeline - glue team please fill this in
+        return True
+    else:
+        return False
+
+
+#helper function for invoke_next_pipeline()
+#checks to see if all the perturbation queries have been finished
+def check_all_queries_made(_id):
+    #this looks inside the alternatives collection
+    perturbed_trips = query_alternatives(u_id) #looks into the database
+    for perturbed_u_id in perturbed_trips:
+        if perturbed_trips[perturbed_u_id] == None:
+            return False
+    return True
+
+
+def store_alternative_trips(tripObj):
+    # store populated tripObj with _id (concatenated trip id and user id)
+    db = get_alternative_trips_db()
+    _id = tripObj.get_id()
+    json = E_Mission_Trip.trip_to_json(tripObj)
+    db.insert_one({_id : json})
+
+def get_alternative_trips(_id):
+    # User Utility Pipeline calls this to get alternatve trips for one original trip (_id)
+    # db = get_alternative_trips_db()
+    # _id = tripObj.get_id()
+    # return db.find(_id)
+    return [Trip(jsonStr) for jsonStr in get_alternative_trips_db().find({'_id' : _id})].__iter__()
+
+def store_perturbed_trips(tripObj):
+    # store populated tripObj with _id (concatenated trip id and user id)
+    db = get_perturbed_trips_db()
+    _id = tripObj.get_id()
+    json = E_Mission_Trip.trip_to_json(tripObj)
+    db.insert_one({_id : json})
+
+def get_perturbed_trips(_id):
+    # User Utility Pipeline calls this to get alternatve trips for one original trip (_id)
+    # db = get_perturbed_trips_db()
+    # _id = tripObj.get_id()
+    # return db.find(_id)
+    return [Trip(jsonStr) for jsonStr in get_perturbed_trips_db().find({'_id' : _id})].__iter__()
+
+
+
+
+# Invoked to find the choices user had when making their trip
+#Jeff - what is this code doing? it seems to do nothing
+def calc_alternative_trips(trip_iterator):
+    #schedule_queries(trip_iterator) 
```

Comment:
@jeffdh5, @joshzarrabi, @shanthistyle, I am completely confused about how this is supposed to work. The _only_ function that is invoked from the pipeline is `calc_alternative_trips`, which does nothing. How is the pipeline supposed to work? What did your tests actually test?